modules/core/src/main/scala/iota/internal/CopKFunctionKMacros.scala

```diff
@@ -77,34 +77,19 @@ final class CopKFunctionKMacros(val c: Context) {
     arrs: List[Tree]
   ): Tree = {
 
-    val namedArrs = arrs.zipWithIndex.map { case (arr, i) =>
-      (TermName(s"arr$i"), arr, i) }
-
-    val defs = namedArrs.map { case (n, arr, _) =>
-      q"private[this] val $n = $arr.asInstanceOf[_root_.cats.arrow.FunctionK[Any, $G]]" }
-
-    val cases = namedArrs.map { case (n, _, i) =>
-      cq"$i => $n(ca.value)" }
-
-    val toStringValue = s"CopKFunctionK[$F, $G]<<generated>>"
-
-    def symbol(tpe: Type): Symbol = tpe match {
-      case TypeRef(_, sym, Nil) => sym
-      case _                    => tpe.typeSymbol
+    val handlers = arrs.zipWithIndex.map { case (arr, i) =>
+      val name = TermName(s"arr$i")
+      val pre = q"private[this] val $name = $arr.asInstanceOf[_root_.cats.arrow.FunctionK[Any, $G]]"
+      val handler = (fa: TermName) => q"$name($fa.value)"
+      (pre, handler)
     }
 
-    q"""
-    new _root_.iota.CopKFunctionK[$F, $G] {
-      ..$defs
-      override def apply[A](ca: ${symbol(F)}[A]): ${symbol(G)}[A] =
-        (ca.index: @_root_.scala.annotation.switch) match {
-          case ..$cases
-          case i => throw new _root_.java.lang.Exception(
-            s"iota internal error: index " + i + " out of bounds for " + this)
-        }
-      override def toString: String = $toStringValue
-    }
-    """
+    tb.makeFastFunctionK(
+      TypeName(c.freshName("CopKFunctionK")),
+      F, G,
+      preamble = handlers.map(_._1),
+      toIndex  = fa => q"$fa.index",
```

Comment:
If `index` is the actual name for the synthetic field added to the ADTs we need to obscure or namespace it sine it's a very common property name.